appinventor/blocklyeditor/src/exportBlocksImage.js

```diff
@@ -517,6 +517,75 @@ Blockly.exportBlockAsPng = function(block) {
   });
 };
 
+/**
+ * Extracts the block types from the given XML.
+ * @param {Document} xml - The XML document containing blocks.
+ * @returns {Array<string>} An array of block types.
+ */
+function extractBlockTypes(xml) {
+  var blockTypes = [];
+  var blocks = xml.getElementsByTagName('block');
+  for (var i = 0; i < blocks.length; i++) {
+    var mutation = blocks[i].getElementsByTagName('mutation')[0];
```

Comment:
Note that `getElementsByTagName` gets all descendants, so you could just do `xml.getElementsByTagName('mutation')` to get all of the mutations. Otherwise, this code ends up being quadratic in its runtime since for each block you may be looking up mutations on its children, grandchildren, etc. not just its own mutation.